ts/packages/ts/src/ConfigSchema.ts

```diff
@@ -1,463 +1,496 @@
+    return;
+  }
 
-  static TYPES = ['json'];
-  static EXT = ConfigSchema.TYPES.map((type) => `${ConfigSchema.CFGU.EXT}.${type}`);
+  if (!cfgu.options.some((option) => option === value)) {
+    throw new ConfigError(
+      'invalid config value',
+      `value "${value}" must be one of ${_.map(cfgu.options, (option) => `"${option}"`).join(',')}`,
+    );
+  }
+};
+const cfguValueTypeValidator = (cfgu: Cfgu, value: string) => {
+  const { type } = cfgu;
 
-  public readonly type: ConfigSchemaType;
+  const validate = CFGU_VALUE_TYPE_VALIDATORS[type];
+  if (!validate) {
+    throw new ConfigError(
+      'invalid type property',
+      `type "${type}" is not yet supported in this SDK. For the time being, please utilize the "String" type. We'd greatly appreciate it if you could open an issue regarding this at https://github.com/configu/configu/issues/new/choose so we can address it in future updates.`,
+    );
+  }
 
-  constructor(public readonly path: string) {
-    const scopeLocation = [`ConfigSchema`, `constructor`];
-    const splittedPath = path.split('.');
+  const isValid = validate({ ...cfgu, value });
+  if (isValid) {
+    return;
+  }
 
-    const fileExt = splittedPath.pop();
-    if (!fileExt || !ConfigSchema.TYPES.includes(fileExt)) {
-      throw new Error(
-        ERR(`invalid path "${path}"`, {
-          location: scopeLocation,
-          suggestion: `path extension must be ${ConfigSchema.EXT.join('|')}`,
-        }),
-      );
+  let hint = `value "${value}" must be of type "${type}"`;
+  // if (cfgu?.options) {
+  //   hint = `value "${value}" must be one of ${_.map(cfgu.options, (option) => `"${option}"`).join(',')}`;
+  // } else
+  if (type === 'RegEx') {
+    hint = `value "${value}" must match the pattern "${cfgu.pattern}"`;
+  } else if (type === 'JSONSchema') {
+    hint = `value "${value}" must match the schema "${cfgu.schema}"`;
+  }
+
+  throw new ConfigError('invalid config value', hint);
+};
+const cfguStructureValidator = (cfgu: Cfgu) => {
+  try {
+    Convert.cfguToJson(cfgu);
+  } catch (error) {
+    throw new ConfigError('invalid cfgu structure', error.message);
+  }
+
+  if (cfgu.type === 'RegEx' && !cfgu.pattern) {
+    throw new ConfigError('invalid type property', `type "${cfgu.type}" must come with a pattern property`);
+  }
+
+  if (cfgu.type === 'JSONSchema' && !cfgu.schema) {
+    throw new ConfigError('invalid type property', `type "${cfgu.type}" must come with a schema property`);
+  }
+
+  if (cfgu.options) {
+    const reason = 'invalid options property';
+    if (_.isEmpty(cfgu.options)) {
+      throw new ConfigError(reason, `options mustn't be empty if set`);
     }
-    this.type = fileExt as ConfigSchemaType;
+    if (cfgu.template) {
+      throw new ConfigError(reason, `options mustn't set together with template properties`);
+    }
+    cfgu.options.forEach((option, idx) => {
+      // todo: rethink on this restriction - what if the user sees empty string as a valid value?
```

Comment:
Fair point - the main concern was (as far as I remember) that the `UpsertCommand` treats upserting an empty string as deleting a key, which conflicts with empty strings being valid values. Worth re-thinking for sure, but may not be as simple as just allowing empty strings in `options`.

Reply:
Totally get your point, @jamesL92. I've left a comment in the code to keep your thoughts in mind for later. Thanks a lot for highlighting this!